syntax = "proto3";
package payload.manager.v1;

import "payload/manager/v1/placement.proto";
import "payload/manager/v1/policy.proto";
import "payload/manager/v1/types.proto";
import "payload/manager/v1/id.proto";

/*
  BLOCKING policy waits until requested durability
  or definitive failure.
*/
message PromoteRequest {
  PayloadID id = 1;
  Tier target_tier = 2;
  PromotionPolicy policy = 3;
}

message PromoteResponse {
  PayloadDescriptor payload_descriptor = 1;
}

message SpillRequest {
  repeated PayloadID ids = 1;
  SpillPolicy policy = 2;
  bool wait_for_leases = 3;
  bool fsync = 4;
}

message SpillResponse {
  message Result {
    PayloadID id = 1;
    bool ok = 2;
    string error_message = 3;
    PayloadDescriptor payload_descriptor = 4;
  }
  repeated Result results = 1;
}

/*
  Hint only â€” no guarantee.
*/
message PrefetchRequest {
  PayloadID id = 1;
  Tier target_tier = 2;
}

/*
  Stronger than eviction policy but still best-effort.
*/
message PinRequest {
  PayloadID id = 1;
  uint64 duration_ms = 2;
}
